Frameworks/CoreGraphics/CGImage.mm

```diff
@@ -21,981 +21,660 @@
-                b /= 255;
-                a *= maskAA;
-                a /= 255;
-
-                *newImageData = r | (g << 8) | (b << 16) | (a << 24);
-
-                newImageData++;
-                srcRow++;
-                srcMaskX += incX;
-            }
-        }
-
-        image->Backing()->ReleaseImageData();
-        mask->Backing()->ReleaseImageData();
-    }
-
-    return (CGImageRef)newImage;
-}
+CGImageRef CGImageMaskCreate(size_t width, size_t height,
+                             size_t bitsPerComponent, size_t bitsPerPixel,
+                             size_t bytesPerRow, CGDataProviderRef provider,
+                             const CGFloat *decode, bool shouldInterpolate) {
+  RETURN_NULL_IF(((bitsPerComponent == 8) && (bitsPerPixel == 32)));
 
-/**
- @Status Caveat
- @Notes Only 32bpp RGBA source format supported. Returns an 8bpp grayscale alpha mask one-time
-        copy of source bitmap.
-*/
-CGImageRef CGImageMaskCreate(size_t width,
-                             size_t height,
-                             size_t bitsPerComponent,
-                             size_t bitsPerPixel,
-                             size_t bytesPerRow,
-                             CGDataProviderRef provider,
-                             const CGFloat* decode,
-                             bool shouldInterpolate) {
-    FAIL_FAST_HR_IF_FALSE(E_UNEXPECTED, ((bitsPerComponent == 8) && (bitsPerPixel == 32)));
-    FAIL_FAST_HR_IF_FALSE_MSG(E_INVALIDARG,
-                              ((provider == nullptr) || [(NSObject*)provider isKindOfClass:[NSData class]]),
-                              "CGDataProviderRef does not derive from NSData!");
-
-    NSData* dataProvider = (__bridge NSData*)provider;
-    char* pData = (char*)[dataProvider bytes];
-    size_t dataLen = (size_t)[dataProvider length];
-
-    //  Create an 8-bit mask from the data
-    __CGSurfaceInfo surfaceInfo = _CGSurfaceInfoInit(width, height, _ColorGrayscale);
-
-    CGImageRef newImage = new CGBitmapImage(surfaceInfo);
-    char* pNewImage = (char*)newImage->Backing()->LockImageData();
-
-    int pixOut = 0;
-    int outImageBytesPerRow = newImage->Backing()->BytesPerRow();
-
-    for (unsigned y = 0; y < height; y++) {
-        char* rowIn = &pData[y * bytesPerRow];
-        char* rowOut = &pNewImage[pixOut];
-
-        for (unsigned x = 0; x < width; x++) {
-            BYTE r = *rowIn++;
-            BYTE g = *rowIn++;
-            BYTE b = *rowIn++;
-            BYTE a = *rowIn++;
-
-            BYTE alphaOut = (r + g + b) * a / 255 / 3;
-            *rowOut++ = 255 - alphaOut;
-        }
-
-        pixOut += outImageBytesPerRow;
-    }
+  RETURN_NULL_IF(((provider == nullptr) ||
+                  ![(NSObject *)provider isKindOfClass:[NSData class]]));
 
-    newImage->Backing()->ReleaseImageData();
+  NSData *dataProvider = (__bridge NSData *)provider;
 
-    return (CGImageRef)newImage;
-}
+  unsigned char *data = (unsigned char *)[dataProvider bytes];
```

Comment:
given how it's passed to WIC, leave it as is.

---
In reply to: [84141914](https://github.com/Microsoft/WinObjC/pull/1185#discussion_r84141914) [](ancestors = 84141914)